lib/midcom/helper/imagepopup/handler/upload.php

```diff
@@ -0,0 +1,151 @@
+<?php 
+/**
+ * @author lukasz chalat
+ * @package midcom.helper.imagepopup
+ * @copyright The Midgard Project, http://www.midgard-project.org
+ * @license http://www.gnu.org/licenses/lgpl.html GNU Lesser General Public License
+ */
+
+/**
+ * This handler get uploaded file and save it in database and write to file.
+ *
+ * @package midcom.helper.imagepopup
+ */
+
+class midcom_helper_imagepopup_handler_upload extends midcom_baseclasses_components_handler
+{
+    /**
+     * @param mixed $handler_id The ID of the handler.
+     * @param array $args The argument list.
+     * @param array &$data The local request data.
+     */
+    public function _handler_upload($handler_id, array $args, array &$data)
+    {
+        // Get the file
+        reset ($_FILES);
+        $temp = current($_FILES);
+        
+        $accepted_origin = midcom::get()->get_host_name();
+		
+        if(is_uploaded_file($temp['tmp_name']))
+        {
+			         if (isset($_SERVER['HTTP_ORIGIN']))
+			         {
+				            // same-origin requests won't set an origin. If the origin is set, it must be valid.
+				            if (in_array($_SERVER['HTTP_ORIGIN'], $accepted_origin))
+				            {
+				                midcom::get()->header('Access-Control-Allow-Origin: ' . $_SERVER['HTTP_ORIGIN']);
+				            }
+				            else
+				            {
+					               midcom::get()->header("HTTP/1.0 403 Origin Denied");
+					               return;
+				            }
+            }
+		
+            if (preg_match("/([^\w\s\d\-_~,;:\[\]\(\).])|([\.]{2,})/", $temp['name']))
+			         {
+				            midcom::get()->header("HTTP/1.0 500 Invalid file name.");
+				            return;
+			         }
+		
+			         if (!in_array(strtolower(pathinfo($temp['name'], PATHINFO_EXTENSION)), array("gif", "jpg", "png")))
+			         {
+				            midcom::get()->header("HTTP/1.0 500 Invalid extension.");
+				            return;
+			         }
+        }    		
+       
+        // Get the data
+		      $temp_name = $temp['tmp_name'];
+		      $mimetype = $temp['type'];
+				    $parentguid = $args[0];
+		
+		      // Set modified filename 
+		      $filename = $this->get_modify_filename($temp['name']);
+		
+		      // Insert the image into database
+		      $attachment = $this->insert_database($filename, $mimetype, $parentguid);
+        
+        // Get target to write the file
+        $target = $this->get_data_from_database($filename, $parentguid);
+        
+        // Write the file
+        $this->write_the_file($temp_name, $target);
+        
+        // Make a response for uploadImage() function
+        $location = midcom_db_attachment::get_url($attachment);
+        
+        // Return image location as JSON
+		      return new midcom_response_json(array('location' => $location));
+    }
+	
+    private function get_modify_filename($filename)
+	   {
+		      $pieces = explode('.', $filename);
```

Comment:
We should make sure that the filename is URL-safe (see `midcom_db_attachment::safe_filename()`). The query part needs to be adapted accordingly